library/general/src/lib/yast2/popup.rb

```diff
@@ -0,0 +1,318 @@
+        case richtext
+        when true
+          HBox(
+            VSpacing(RICHTEXT_HEIGHT),
+            VBox(
+              HSpacing(RICHTEXT_WIDTH),
+              RichText(message)
+            )
+          )
+        when false
+          Left(Label(message))
+        when :auto
+          if message.lines.size >= LINES_THRESHOLD
+            message_widget(plain_to_richtext(message), true)
+          else
+            message_widget(message, false)
+          end
+        else
+          raise ArgumentError, "Invalid richtext paramter #{richtext.inspect}"
+        end
+      end
+
+      def body(headline, message, richtext, timeout)
+        VBox(
+          VSpacing(0.4),
+          *headline_widgets(headline),
+          message_widget(message, richtext),
+          VSpacing(0.2),
+          timeout_widget(timeout)
+        )
+      end
+
+      def content(body, buttons)
+        HBox(
+          HSpacing(1),
+          VBox(
+            VSpacing(0.2),
+            body,
+            VSpacing(1),
+            button_box(buttons),
+            VSpacing(0.2)
+          ),
+          HSpacing(1)
+        )
+      end
+
+      def button_box(buttons)
+        push_buttons = buttons.map do |id, label|
+          # lets auto detect options for button box
+          opt = case id
+          when :ok, :yes, :continue
+            Opt(:key_F10, :okButton)
+          when :cancel, :no
+            Opt(:key_F9, :cancelButton)
+          else
+            Opt(:customButton)
+          end
+
+          PushButton(Id(id), opt, label)
+        end
+
+        return Empty() if push_buttons.empty?
+
+        # relax sanity check as there can be situation where is
+        # e.g. OK and details with `show(text, details: text2)`
+        ButtonBox(Opt(:relaxSanityCheck), *push_buttons)
+      end
+
+      def event_loop(content, focus, timeout, details, style)
+        res = Yast::UI.OpenDialog(dialog_options(style), content)
+        remaining_time = timeout
+        raise "Failed to open dialog, see logs." unless res
+        begin
+          Yast::UI.SetFocus(focus)
+          loop do
+            res = timeout > 0 ? Yast::UI.TimeoutUserInput(1000) : Yast::UI.UserInput
+            remaining_time -= 1
+            res = handle_event(res, details, remaining_time, focus)
+            return res if res
+          end
+        ensure
+          Yast::UI.CloseDialog
+        end
+      end
+
+      def dialog_options(style)
+        case style
+        when :notice
+          Opt()
+        when :important
+          Opt(:infocolor)
+        when :warning
+          Opt(:warncolor)
+        else
+          raise ArgumentError, "Invalid style parameter #{style.inspect}"
+        end
+      end
+
+      def handle_event(res, details, remaining_time, focus)
+        case res
+        when :details
+          show(details)
+          nil
+        when :timeout
+          if remaining_time <= 0
+            :focus
```

Comment:
Hmmm

Reply:
ah, good catch, it should be `focus` not symbol